### PR TITLE
GH-102613: Fast recursive globbing in `pathlib.Path.glob()`

### PR DESCRIPTION
**Pull request by barneygale**
This PR introduces a 'walk-and-match' strategy for handling glob patterns that include a non-terminal `**` wildcard, such as `**/*.py`. For this example, the previous implementation recursively walked directories using `os.scandir()` when it expanded the `**` component, and then **scanned those same directories again** when expanded the `*.py` component. This is wasteful.

In the new implementation, any components following a `**` wildcard are used to build a `re.Pattern` object, which is used to filter the results of the recursive walk. A pattern like `**/*.py` uses half the number of `os.scandir()` calls; a pattern like `**/*/*.py` a third, etc.

This new algorithm does not apply if either:

1. The *follow_symlinks* argument is set to `None` (its default), or
2. The pattern contains `..` components.

In these cases we fall back to the old implementation.

This PR also replaces selector classes with selector functions. These generators directly yield results rather calling through to their successors. A new internal `Path._glob()` method takes care to chain these generators together, which simplifies the lazy algorithm and slightly improves performance. It should also be easier to understand and maintain.

Performance for the original #102613 repro case, with 400 nested `a/` directories, and matching treatment of symlinks and hidden files:

```shell
$ ../python -m timeit -s 'import glob' 'print(glob.glob("**/*", recursive=True, include_hidden=True))'
5 loops, best of 5: 66.2 msec per loop
$ ../python -m timeit -s 'from pathlib import Path' 'print(list(Path(".").rglob("**/*", follow_symlinks=True)))'
10 loops, best of 5: 22.7 msec per loop  # before this PR
10 loops, best of 5: 16.5 msec per loop  # after this PR
```

These results were from an SSD. The improvement will be greater for slow storage (e.g. network-mounted volumes).

* Issue: gh-102613

### CODE REVIEW

**Comment by barneygale** [comment]
@zooba here's the promised walk-and-match implementation!

File: Lib/pathlib.py [hunk 1]
```diff
@@ -1056,51 +977,112 @@ def _scandir(self):
...
+                pass
+            elif part == '..':
+                paths = (path._make_child_relpath('..') for path in paths)
+            elif part == '**':
+                filter_paths = False
+                if filter_paths_supported:
+                    # Consume remaining path components, except trailing slash.
+                    while part_idx < len(pattern_parts) and pattern_parts[part_idx] != '':
+                        filter_paths = True
+                        part_idx += 1
```
Comment (zooba) [inline Lib/pathlib.py]: The comment seems to suggest that we only get an empty `pattern_parts[i]` after a trailing slash, which presumably can only be the last element? If so, we should be able to simplify this down to:

```
filter_paths = True
part_idx = len(pattern_parts)
if pattern_parts[-1] == '':
    part_idx -= 1
```

`filter_paths` is guaranteed to be updated right now (we already know the current part is `'**'`), and so the whole loop is unnecessary if we know how `''` gets in there.

If we can get a `''` part some other way, that probably needs special handling. But I think we normalise paths before they become Paths, right?

File: Lib/pathlib.py [hunk 2]
```diff
@@ -1056,51 +977,112 @@ def _scandir(self):
...
-        selector = _make_selector(("**",) + tuple(pattern_parts), self._flavour, case_sensitive)
-        for p in selector.select_from(self, follow_symlinks):
-            yield p
+
+        if case_sensitive is None:
+            # TODO: evaluate case-sensitivity of each directory in _select_children().
+            case_sensitive = _is_case_sensitive(self._flavour)
+
+        # If symlinks are handled consistently, and the pattern does not
+        # contain '..' components, then we can use a 'walk-and-match' strategy
```
Comment (zooba) [inline Lib/pathlib.py:1055]: Can the pattern contain `..` components _before_ the first `**` (but possibly after other wildcards)?

File: Lib/pathlib.py (hunk 1 above)
Comment (zooba) [reply Lib/pathlib.py]: And perhaps we can just move the `dir_only` calculation into this `if` block so that it looks more like:

```
filter_paths = True
part_idx = len(pattern_parts)
dir_only = pattern_parts[-1] == ''
```

and the `else` gets the other case?

At this point, it looks like we're down to a single `_select_recursive` call shared between the two code paths, so perhaps the "we can filter" vs "we can't filter" cases make more sense as a higher level condition (i.e. at line 1055)? I'll leave that one up to you.

File: Lib/pathlib.py (hunk 2 above)
Comment (zooba) [reply Lib/pathlib.py:1055]: Ah, reading below it looks like we use the whole pattern to filter, not just the part remaining after the `**`.

File: Lib/pathlib.py [hunk 3]
```diff
@@ -1056,51 +977,112 @@ def _scandir(self):
...
+    def _glob(self, pattern, case_sensitive, follow_symlinks):
+        path_pattern = self.with_segments(pattern)
+        if path_pattern.drive or path_pattern.root:
             raise NotImplementedError("Non-relative patterns are unsupported")
-        if pattern and pattern[-1] in (self._flavour.sep, self._flavour.altsep):
+        elif not path_pattern._tail:
+            raise ValueError("Unacceptable pattern: {!r}".format(pattern))
+
+        pattern_parts = list(path_pattern._tail)
+        if pattern[-1] in (self._flavour.sep, self._flavour.altsep):
```
Comment (zooba) [inline Lib/pathlib.py:1043]: Have we validated that `pattern` is not empty here? The old code also checked `if pattern and ...`

**Review by zooba — COMMENTED** [review]
This seems fine, there's potentially a few simplifications, but it looks like a great improvement over the existing code.

Do we need any new tests to specifically trigger anything that behaves differently?

File: Lib/pathlib.py (hunk 2 above)
Comment (barneygale) [reply Lib/pathlib.py:1055]: > Can the pattern contain `..` components _before_ the first `**` (but possibly after other wildcards)?

It can, and technically we don't need to use the slow algorithm if the `..` occurs before the `**`. But I fear the algorithm will become too hard to follow when evaluating a pattern like `**/../**/*`, which would use the slow and then the fast algorithm.

> Ah, reading below it looks like we use the whole pattern to filter, not just the part remaining after the `**`.

That's right. If we use only the bit after `**`, we have to put a `relative_to()` call in there, I think. There may be a way to do it by removing the `self._lines` prefix from `pattern_lines`

File: Lib/pathlib.py (hunk 3 above)
Comment (barneygale) [reply Lib/pathlib.py:1043]: If an empty pattern is supplied, the `elif not path_pattern._tail:` check will fail

File: Lib/pathlib.py (hunk 1 above)
Comment (barneygale) [reply Lib/pathlib.py]: I actually had code identical to yours at one stage!

> `filter_paths` is guaranteed to be updated right now (we already know the current part is `'**'`), and so the whole loop is unnecessary if we know how `''` gets in there.

Not quite! `filter_paths` is only turned on if there's at least one non-empty part after the `**`. There's no point doing a regex match for a pattern like `a/**/`. 

Still, I'll work on this code to see if there's a better way.

File: Lib/pathlib.py (hunk 2 above)
Comment (barneygale) [reply Lib/pathlib.py:1055]: I've found a way to exclude `self` from the matching (see [064efdb](https://github.com/python/cpython/pull/104512/commits/064efdb452f1e0db040d1bda2b2faa9ef57df8c2)) but it's harder to exclude earlier parts of the pattern - it requires extra bookkeeping that negates the benefit.

File: Lib/pathlib.py (hunk 1 above)
Comment (barneygale) [reply Lib/pathlib.py]: Solved, I think!

**Review by zooba — APPROVED** [review]
